src/Bitmap_cubical_complex/include/gudhi/Persistence_on_rectangle.h

```diff
@@ -0,0 +1,595 @@
+ *    Author(s): Marc Glisse
+ *
+ *    Copyright (C) 2023 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+//  ds_find_set_ is inspired from code in Boost.Graph that is
+//
+//  (C) Copyright Jeremy Siek 2004
+//  Distributed under the Boost Software License, Version 1.0. (See
+//  accompanying file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+
+#ifndef PERSISTENCE_ON_RECTANGLE_H
+#define PERSISTENCE_ON_RECTANGLE_H
+
+#include <gudhi/Debug_utils.h>
+#ifdef GUDHI_DETAILED_TIMES
+ #include <gudhi/Clock.h>
+#endif
+
+#include <boost/range/adaptor/reversed.hpp>
+
+#ifdef GUDHI_USE_TBB
+ #include <tbb/parallel_sort.h>
+#endif
+
+#ifdef DEBUG_TRACES
+ #include <iostream>
+#endif
+#include <vector>
+#include <memory>
+#include <algorithm>
+#include <stdexcept>
+#include <cstddef>
+
+namespace Gudhi::cubical_complex {
+
+// When building a cubical complex from top-dimensional cells, there are
+// normally more vertices than input top cells ((x+1)*(y+1) instead of x*y).
+// However, the top cells along the boundary turn out to be collapsible, and we
+// only need to work with (x-1)*(y-1) vertices.
+template <class Filtration_value, class Index = std::size_t, bool output_index = false>
+struct Persistence_on_rectangle {
+  // If we want to save space, we don't have to store the redundant 'first'
+  // field in T_with_index. However, it would slow down the primal pass.
+  struct T_with_index {
+    Filtration_value first; Index second;
+    T_with_index() = default;
+    T_with_index(Filtration_value f, Index i) : first(f), second(i) {}
+    bool operator<(T_with_index const& other) const { return std::tie(first, second) < std::tie(other.first, other.second); }
+    Index out() const { return second; }
+  };
+  // Don't store the index if we don't want to output it.
+  struct T_no_index {
+    Filtration_value first;
+    T_no_index() = default;
+    T_no_index(Filtration_value f, Index) : first(f) {}
+    bool operator<(T_no_index const& other) const { return first < other.first; }
+    Filtration_value out() const { return first; }
+  };
+  typedef std::conditional_t<output_index, T_with_index, T_no_index> T;
+
+  Filtration_value const* input_p;
+  Filtration_value input(Index i) const { return input_p[i]; }
+
+  // size_* counts the number of vertices in each direction.
+  Index size_x, size_y, input_size;
+  // The square i + dy is right above i.
+  Index dy;
+
+  // Squares keep their index from the input.
+  // Vertices have the same index as the square at their bottom left (smaller x and y)
+  // Store the filtration value of vertices that could be critical. We could store them
+  // in some map, or recompute them on demand, but this strongly affects performance.
+  std::unique_ptr<T[]> data_v_;
+  T& data_vertex(Index i){ return data_v_[i]; }
+  T data_vertex(Index i) const { return data_v_[i]; }
```

Comment:
As the access levels of `data_v_[i]` and `data_vertex(i)` are the same, are those getter function not useless? (see also below for ds_parents_*) Even if later, say, `ds_parent_v_` becomes private, would the method `ds_parent_vertex` really be useful as public method?

Reply:
The goal wasn't to have getters for a private member, the goal was to abstract away how this data is stored. At various points it was stored in a map, or accessed as storage[i/2].filt, or...
It is true that it ends up as a really trivial wrapper...

Reply:
Ok, makes sense.